flight/PiOS/Common/pios_px4flow.c

```diff
@@ -0,0 +1,286 @@
+ */
+static int32_t PIOS_PX4Flow_Validate(struct px4flow_dev *dev)
+{
+	if (dev == NULL)
+		return -1;
+	if (dev->magic != PIOS_PX4FLOW_DEV_MAGIC)
+		return -2;
+	if (dev->i2c_id == 0)
+		return -3;
+	return 0;
+}
+
+/**
+ * @brief Initialize the PX4Flow optical flow sensor.
+ * @return 0 on success
+ */
+int32_t PIOS_PX4Flow_Init(const struct pios_px4flow_cfg *cfg, const uint32_t i2c_id)
+{
+	dev = (struct px4flow_dev *) PIOS_PX4Flow_alloc();
+	if (dev == NULL)
+		return -1;
+
+	dev->cfg = cfg;
+	dev->i2c_id = i2c_id;
+	PIOS_PX4Flow_SetRotation(cfg->rotation);
+
+	if (PIOS_PX4Flow_Config(cfg) != 0)
+		return -2;
+
+	PIOS_SENSORS_Register(PIOS_SENSOR_OPTICAL_FLOW, dev->optical_flow_queue);
+	PIOS_SENSORS_Register(PIOS_SENSOR_RANGEFINDER, dev->rangefinder_queue);
+
+	dev->task = PIOS_Thread_Create(PIOS_PX4Flow_Task, "pios_px4flow", PX4FLOW_TASK_STACK_BYTES, NULL, PX4FLOW_TASK_PRIORITY);
+
+	PIOS_Assert(dev->task != NULL);
+
+	return 0;
+}
+
+/**
+ * @brief Updates the PX4Flow's orientation.
+ * @returns 0 for success or -1 for failure
+ */
+int32_t PIOS_PX4Flow_SetRotation(const struct Rotation rotation)
+{
+	if (PIOS_PX4Flow_Validate(dev) != 0)
+		return -1;
+
+	float rpy_R[3]  = {rotation.roll_D100/100.0f * DEG2RAD, rotation.pitch_D100/100.0f * DEG2RAD, rotation.yaw_D100/100.0f * DEG2RAD};
+
+	Euler2R(rpy_R, dev->Rsb);
+
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the PX4Flow optical flow sensor
+ * \return none
+ * \param[in] PIOS_PX4Flow_ConfigTypeDef struct to be used to configure sensor.
+ *
+ */
+static int32_t PIOS_PX4Flow_Config(const struct pios_px4flow_cfg * cfg)
+{
+	// This function intentionally left blank.
+
+	return 0;
+}
+
+/**
+ * @brief Read current optical flow and sonar rangefinding measurements
+ * \param[out] optical_flow_data Struct containing translational x-, y-, and z-axis velocity
+ * \param[out] rangefinder_data Struct containing sonar ranging measurement along z-axis
+ * \return 0 for success or -1 for failure
+ */
+static int32_t PIOS_PX4Flow_ReadData(struct pios_sensor_optical_flow_data *optical_flow_data, struct pios_sensor_rangefinder_data *rangefinder_data)
+{
+	if (PIOS_PX4Flow_Validate(dev) != 0) {
+		return -1;
+	}
+
+	/* don't use PIOS_PX4Flow_Read and PIOS_PX4Flow_Write here because the task could be
+	 * switched out of context in between which would give the sensor less time to capture
+	 * the next sample.
+	 */
+	uint8_t addr_read[] = {
+		PIOS_PX4FLOW_FRAMECOUNTER_LSB
+	};
+
+	/* Optical flow structs */
+	struct I2C_Frame
+	{
+		uint16_t frame_count; // counts created I2C frames [#frames]
+		int16_t pixel_flow_x_sum_px10; // latest x flow measurement in pixels*10 [pixels]
+		int16_t pixel_flow_y_sum_px10; // latest y flow measurement in pixels*10 [pixels]
+		int16_t flow_comp_x_m1000; // x velocity*1000 [meters/sec]
+		int16_t flow_comp_y_m1000; // y velocity*1000 [meters/sec]
+		int16_t qual; // Optical flow solution quality [0: bad, 255: maximum quality]
```

Comment:
So you're suggesting that we could use this field as an error check? If the upper byte is non-zero, we throw out the entire packet?

That makes a lot of sense to me, but it depends on PX4Flow intentionally writing 0 there. I suspect/hope they do, since otherwise their description of the field is off...